### PR TITLE
Updated to PVR API v4.2.0 (implemented IsRealTimeStream)

### PR DESCRIPTION
**Pull request by ksooo**
Implementation should basically do the same as pvr.vdr.vnsi: https://github.com/kodi-pvr/pvr.vdr.vnsi/commit/3b94c58114f93c42e38741fb1dd403858bfe04b2#diff-34f48ad91ac6c202ac60b0348ae90e30R862

@Jalle19 good to go?

### CODE REVIEW

File: src/client.cpp [hunk 1]
```diff
@@ -301,6 +301,39 @@ bool IsTimeshifting(void)
...
 }
 
+static time_t ConvertMusecsToTime(int64_t musecs)
+{
+  // tvheadend reports microseconds for timeshifting values,
+  // Kodi expects it to be an absolute UNIX timestamp
+  return time(NULL) - static_cast<time_t>(static_cast<double>(musecs / 1000000));
+}
+
+time_t GetPlayingTime()
```
Comment (FernetMenta) [inline src/client.cpp]: this does not look thread safe. at least I don't see it

Comment (Jalle19) [inline src/client.cpp]: What do you mean? The code has simply been moved, it's been like this for quite some time.

Comment (FernetMenta) [inline src/client.cpp]: I did not say that it broke but that it does not look right. I have never looked into this before. Those methods here are called be gui but as far as I can see they go unprotected into demuxer.

Comment (ksooo) [inline src/client.cpp]: @FernetMenta you are right. Not introduced by this PR (at least not for the "old" functions), but should be fixed. Will do.

**Comment by ksooo** [comment]
@Jalle19 looked almost first time into class CHTSPDemuxer. Is there any logic when member access is to protect by m_mutex? Cannot make any sense out of what I'm seeing there.

**Comment by Jalle19** [comment]
There is some logic but I think some parts need some work. All the parse methods are called with mutex protection from ProcessMessage(). m_subscription is thread-safe. It's the rest I'm not really sure about.

**Comment by ksooo** [comment]
@FernetMenta  Okay now? (At least the stuff touched by this PR?)

**Comment by FernetMenta** [comment]
@ksooo not sure. where is the protection?

**Comment by ksooo** [comment]
@FernetMenta => https://github.com/ksooo/pvr.hts/commit/a300442439a3df6fe565d5d64f63dac61a4de977#diff-fc67c1aab2e0d4a225199143a1d4be69R236

**Comment by FernetMenta** [comment]
This does not help. GetBufferTimeStart is called by gui thread and accesses an unprotected parameter.

**Comment by ksooo** [comment]
Sure. Sigh. I have the feeling I have to fix more and more stuff others implemented in a non-ideal way... m_dmx_active is accessed without protection "everywhere".

**Comment by FernetMenta** [comment]
what is the story behind m_dxm_active? there's never more than a single demuxer at a given time.

**Comment by Jalle19** [comment]
@FernetMenta yes there are, the active one points to the one that should be sending to Kodi.

@ksooo it should be enough to add mutex locks tp the various Get() methods in the demuxer.

**Comment by FernetMenta** [comment]
@Jalle19 are you playing with demuxers just for fun? kodi never opens more than a single demuxer.

**Comment by Jalle19** [comment]
@FernetMenta it's for an optional feature called predictive tuning.

**Comment by Jalle19** [comment]
@ksooo I'm taking some notes and looking through the demuxer code right now, okay if I handle the threading issues?

**Comment by ksooo** [comment]
Sure. Thx. So, okay to merge this PR now and to address the trhreading stuff in your followup-up PR?

**Comment by Jalle19** [comment]
Yes, let's do that.

File: src/HTSPDemuxer.cpp [hunk 2]
```diff
@@ -231,6 +231,19 @@ void CHTSPDemuxer::SetStreamingProfile(const std::string &profile)
   m_subscription.SetProfile(profile);
 }
 
+bool CHTSPDemuxer::IsRealTimeStream() const
+{
+  CLockObject lock(m_mutex);
```
Comment (Jalle19) [inline src/HTSPDemuxer.cpp]: @ksooo can you drop this before merging?

File: src/Tvheadend.h [hunk 3]
```diff
@@ -273,9 +274,9 @@ class CHTSPDemuxer
   void SetStreamingProfile(const std::string &profile);
 
 private:
-  P8PLATFORM::CMutex                        m_mutex;
+  mutable P8PLATFORM::CMutex              m_mutex;
```
Comment (Jalle19) [inline src/Tvheadend.h]: If you drop the lock you can drop this as well, I'll be re-introducing it in my PR.

Comment (ksooo) [inline src/Tvheadend.h]: sure, will drop both.

**Pull request merged**